lib/grant-types/authorization-code-grant-type.js

```diff
@@ -0,0 +1,210 @@
+
+/**
+ * Module dependencies.
+ */
+
+var AbstractGrantType = require('./abstract-grant-type');
+var InvalidArgumentError = require('../errors/invalid-argument-error');
+var InvalidGrantError = require('../errors/invalid-grant-error');
+var InvalidRequestError = require('../errors/invalid-request-error');
+var Promise = require('bluebird');
+var ServerError = require('../errors/server-error');
+var is = require('../validator/is');
+var util = require('util');
+
+/**
+ * Constructor.
+ */
+
+function AuthorizationCodeGrantType(options) {
+  options = options || {};
+
+  if (!options.model) {
+    throw new InvalidArgumentError('Missing parameter: `model`');
+  }
+
+  if (!options.model.getAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `getAuthorizationCode()`');
+  }
+
+  if (!options.model.revokeAuthorizationCode) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `revokeAuthorizationCode()`');
+  }
+
+  if (!options.model.saveToken) {
+    throw new InvalidArgumentError('Invalid argument: model does not implement `saveToken()`');
+  }
+
+  AbstractGrantType.call(this, options);
+}
+
+/**
+ * Inherit prototype.
+ */
+
+util.inherits(AuthorizationCodeGrantType, AbstractGrantType);
+
+/**
+ * Handle authorization code grant.
+ *
+ * @see https://tools.ietf.org/html/rfc6749#section-4.1.3
+ */
+
+AuthorizationCodeGrantType.prototype.handle = function(request, client) {
+  if (!request) {
+    throw new InvalidArgumentError('Missing parameter: `request`');
+  }
+
+  if (!client) {
+    throw new InvalidArgumentError('Missing parameter: `client`');
+  }
+
+  var scope = this.getScope(request);
```

Comment:
You should not be taking scope from the user input. Access tokens should use the authorization codes's granted scope because it's not guaranteed the user granted access to the scopes requested..

Fixed in my PR at https://github.com/night/node-oauth2-server/commit/9a001a1255849a3d534bbd7fa78fc6d6c1726895#diff-99565ee28a520ab376c88b2fdefffeb6L62